### PR TITLE
feat: add group component

### LINKED ISSUE #73: [backend] Group Component

### PR DESCRIPTION
**Pull request by dkim94**
### What is the issue this pull request related with?

resolved #73 

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.

- [X] Code builds correctly
- [ ] Created tests which fail without the change (if possible)
- [X] All tests passing
- [ ] Extended the README / documentation, if necessary

### Description
temporary add backend group component

💔Thank you!

### CODE REVIEW

File: backend/api_gateway/carpool_request/views.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def __create_carpool_request(request):
         from_location=body['from_location'], 
         to_location=body['to_location'],
         minimum_passenger=body['minimum_passenger'], 
-        rider_id=body['rider_id'],
```
Comment (mss8087) [inline backend/api_gateway/carpool_request/views.py]: use rider_id!

File: backend/api_gateway/group/views.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    if request.method == 'POST':
+        return __create_group(request)
+    else:
+        return HttpResponseNotAllowed(['POST'])
+
+def __create_group(request):
+    body = json.loads(request.body.decode())
+    # TODO: check KeyError
+    command = GroupCreateCommand(
+        riders=body['riders'], from_location=body['from_location'],
```
Comment (mss8087) [inline backend/api_gateway/group/views.py]: use rider_id_list!

File: backend/api_gateway/group/views.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return HttpResponseNotAllowed(['POST'])
+
+def __create_group(request):
+    body = json.loads(request.body.decode())
+    # TODO: check KeyError
+    command = GroupCreateCommand(
+        riders=body['riders'], from_location=body['from_location'], 
+        to_location=body['to_location'])
+
+    result = RedisRpcClient().call(GROUP_CREATE_COMMAND, command)
```
Comment (mss8087) [inline backend/api_gateway/group/views.py]: Refer to carpool_request_service/views.py
result should be preprocessed to json data.

File: backend/common/command/group_create_command.py [hunk 4]
```diff
@@ -2,27 +2,27 @@
 
 GROUP_CREATE_COMMAND = 'command.group_create'
 
-
 class GroupCreateCommand(Command):
 
-    def __init__(self, from_location, to_location, rider_id_list):
+    def __init__(self, riders, from_location, to_location):
```
Comment (mss8087) [inline backend/common/command/group_create_command.py]: use rider_id_list

File: backend/common/command/group_create_command.py [hunk 5]
```diff
@@ -2,27 +2,27 @@
...
 
     @property
     def from_location(self):
         return self._from_location
-
     @property
     def to_location(self):
         return self._to_location
 
-    @property
```
Comment (mss8087) [inline backend/common/command/group_create_command.py]: remain the rider_id_list method

File: backend/common/command/group_create_command.py [hunk 6]
```diff
@@ -2,27 +2,27 @@
...
-    def rider_id_list(self):
-        return self._rider_id_list
-
     def __str__(self):
-        return 'rider_number={},from={},to={}'.format(
-            len(self._rider_id_list), self._from_location, self._to_location)
+        return 'riders={},from_location={},to_location={}'.format(
+            self._riders, self._from_location, self._to_location)
```
Comment (mss8087) [inline backend/common/command/group_create_command.py]: use rider_id_list

File: backend/common/command/tests/test_group_create_command.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+from django.test import TestCase
+
+from backend.common.command.group_create_command import GroupCreateCommand
+
+class GroupCreateCommandTestCase(TestCase):
+
+    def test_properties(self):
+        command = GroupCreateCommand(
+            riders=['rider1', 'rider2', 'rider3', 'rider4'],
```
Comment (mss8087) [inline backend/common/command/tests/test_group_create_command.py]: use rider_id_list

File: backend/common/event/group_created_event.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+from backend.common.event.domain_event import DomainEvent
+
+GROUP_CREATED_EVENT = 'event.group_created'
+
+
+class GroupCreatedEvent(DomainEvent):
+    
+    def __init__(self, riders, from_location, to_location):
+        super().__init__(GROUP_CREATED_EVENT, "v1")
+        self._riders = riders
```
Comment (mss8087) [inline backend/common/event/group_created_event.py]: rider_id_list

File: backend/group_service/group/app/group_application_service.py [hunk 9]
```diff
@@ -1,3 +1,17 @@
-# TODO: implement me :-)
+import logging
+
+from backend.group_service.group.domain.group import Group
+
+logger = logging.getLogger(__name__)
+
 class GroupApplicationService:
-    pass
+    def createGroup(self, riders, from_location, to_location):
```
Comment (mss8087) [inline backend/group_service/group/app/group_application_service.py]: rider_id_list

File: backend/group_service/group/infra/adapter/group_create_command_handler.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from interface import implements 
+from backend.common.messaging.message_handler import MessageHandler
+
+class GroupCreateCommandHandler(implements(MessageHandler)):
+
+    def __init__(self, group_application_service):
+        self.__group_application_service = group_application_service
+
+    def handle(self, message):
+        if message.riders is None or message.from_location is None or \
```
Comment (mss8087) [inline backend/group_service/group/infra/adapter/group_create_command_handler.py]: use rider_id_list

File: backend/api_gateway/group/views.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+    # TODO: handling exception
+    return with_json_response(status=204, data=result)
+
```
Comment (mss8087) [inline backend/api_gateway/group/views.py]: I think we don't need to handle create_group through api_gateaway.
create_group is called with request component only.

Remove this part(?)

File: backend/api_gateway/group/views.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+def __update_group(request):
+    body = json.loads(request.body.decode())
+    # TODO: check KeyError
+    command = GroupUpdateCommand(driver=body['driver'])
+
+    result = RedisRpcClient().call(GROUP_UPDATE_COMMAND, command)
+
+    # TODO: handling exception
+    return with_json_response(status=204, data=result)
```
Comment (mss8087) [inline backend/api_gateway/group/views.py]: driver -> driver_id

refer to api_gateway/carpool_request/views.py
we need to prepross the result of redisrpcclient.call.

File: backend/common/command/group_create_command.py [hunk 13]
```diff
@@ -2,27 +2,27 @@
...
+
+
```
Comment (mss8087) [inline backend/common/command/group_create_command.py]: Use original one.

File: backend/common/event/group_created_event.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+        self._from_location = from_location
+        self._to_location = to_location
+    
+    @property
+    def riders(self):
+        return self._riders
+
```
Comment (mss8087) [inline backend/common/event/group_created_event.py]: User rider_id_list

File: backend/group_service/group/app/group_application_service.py [hunk 15]
```diff
@@ -1,3 +1,17 @@
...
+        return Group.objects.create(riders=riders, driver=None, 
+            from_location=from_location, to_location=to_location)
+
+    def updateGroup(self, groupId, driver):
+        myGroup = Group.objects.get(id=groupId)
+        myGroup.driver = driver
+        return myGroup
```
Comment (mss8087) [inline backend/group_service/group/app/group_application_service.py]: Use update_group, group_id, driver_id
We need to find driver with driver_id => Driver.objects.get(id=driver_id)

File: backend/group_service/group/infra/adapter/group_update_command_handler.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+    def __init__(self, group_application_service):
+        self.__group_application_service = group_application_service
+
+    def handle(self, message):
+        if message.driver is None:
+            raise ValueError("Invalid GroupCreate command parameters")
+
+        self.group_application_service.updateGroup(driver=message.driver)
+            
+
```
Comment (mss8087) [inline backend/group_service/group/infra/adapter/group_update_command_handler.py]: driver => driver_id

File: backend/group_service/group/infra/adapter/tests/test_group_create_command_handler.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+class GroupCreateCommandHandlerTestCase(TestCase):
+
+    def test_when_invalid(self):
+
+        group_create_command_handler = \
+            GroupCreateCommandHandler(group_application_service=None)
+
+        with self.assertRaises(ValueError):
+            group_create_command_handler.handle(
+                GroupCreateCommandHandler(riders=None))
```
Comment (mss8087) [inline backend/group_service/group/infra/adapter/tests/test_group_create_command_handler.py]: rider_id_list

File: backend/group_service/group/infra/adapter/tests/test_group_create_command_handler.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+            self.assertEqual(from_location, FROM_LOCATION)
+            self.assertEqual(to_location, TO_LOCATION)
+
+        mock_group_application_service = \
+            MockGroupApplicationService(assert_func=assert_func)
+        group_create_command_handler = GroupCreateCommandHandler(
+            group_application_service = mock_group_application_service)
+
+        group_create_command_handler.handle(GroupCreateCommandHandler(
+            riders=RIDERS, driver=DRIVER, from_location=FROM_LOCATION, to_location=TO_LOCATION))
```
Comment (mss8087) [inline backend/group_service/group/infra/adapter/tests/test_group_create_command_handler.py]: rider_id_list, driver_id

File: backend/group_service/group/infra/adapter/tests/test_group_update_command_handler.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+        def assert_func(driver):
+            self.assertEqual(driver, DRIVER)
+
+        mock_group_application_service = \
+            MockGroupApplicationService(assert_func=assert_func)
+        group_create_command_handler = GroupCreateCommandHandler(
+            group_application_service = mock_group_application_service)
+
+        group_create_command_handler.handle(GroupCreateCommandHandler(driver=DRIVER))
```
Comment (mss8087) [inline backend/group_service/group/infra/adapter/tests/test_group_update_command_handler.py]: driver_id, rider_id_list

File: backend/api_gateway/group/views.py [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+#     body = json.loads(request.body.decode())
+#     # TODO: check KeyError
+#     command = GroupCreateCommand(
+#         riders=body['riders'], from_location=body['from_location'], 
+#         to_location=body['to_location'])
+
+#     result = RedisRpcClient().call(GROUP_CREATE_COMMAND, command)
+
+#     # TODO: handling exception
+#     return with_json_response(status=204, data=result)
```
Comment (zeroFruit) [inline backend/api_gateway/group/views.py]: Why comment these lines?

Comment (zeroFruit) [reply backend/api_gateway/group/views.py]: Please remove these

File: backend/api_gateway/group/views.py [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+
+def __update_group(request):
+    body = json.loads(request.body.decode())
+    # TODO: check KeyError
+    command = GroupUpdateCommand(driver_id=body['driver_id'])
+
+    result = RedisRpcClient().call(GROUP_UPDATE_COMMAND, command)
+
+    # TODO: handling exception
+    return with_json_response(status=204, data=result)
```
Comment (zeroFruit) [inline backend/api_gateway/group/views.py]: Is these feature tested over ARC or Postman?

File: backend/common/command/group_update_command.py [hunk 22]
```diff
@@ -0,0 +1,18 @@
+from .command import Command
+
+GROUP_UPDATE_COMMAND = 'command.group_update'
+
+class GroupUpdateCommand(Command):
+
+    def __init__(self, driver_id):
```
Comment (zeroFruit) [inline backend/common/command/group_update_command.py]: Maybe in addition to `driver_id`, I think `group_id` also needed

File: backend/group_service/group/app/group_application_service.py [hunk 23]
```diff
@@ -1,3 +1,17 @@
...
+
+logger = logging.getLogger(__name__)
+
 class GroupApplicationService:
-    pass
+    def createGroup(self, rider_id_list, from_location, to_location):
+        return Group.objects.create(rider_id_list=rider_id_list,
+            from_location=from_location, to_location=to_location)
+
+    def updateGroup(self, group_id, driver_id):
```
Comment (zeroFruit) [inline backend/group_service/group/app/group_application_service.py]: please change function name format into `snake_case`

File: backend/group_service/group/infra/adapter/group_create_command_handler.py [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+
+class GroupCreateCommandHandler(implements(MessageHandler)):
+
+    def __init__(self, group_application_service):
+        self.__group_application_service = group_application_service
+
+    def handle(self, message):
+        if message.rider_id_list is None or message.from_location is None or \
+        message.to_location is None or message.cost is None or message.departure is None:
+            raise ValueError("Invalid GroupCreate command parameters")
```
Comment (zeroFruit) [inline backend/group_service/group/infra/adapter/group_create_command_handler.py]:
```python
message.cost is None or message.departure is None
```
This condition will always false. So I think we don't need this logic

File: backend/group_service/group/infra/adapter/tests/test_group_create_command_handler.py [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+        class MockGroupApplicationService:
+
+            def __init__(self, assert_func):
+                self.assert_func = assert_func
+
+            def create(self, rider_id_list, from_location, to_location):
+                self.assert_func(rider_id_list, from_location, to_location)
+
+        RIDER_ID_LIST = ['1', '2', '3', '4']
```
Comment (zeroFruit) [inline backend/group_service/group/infra/adapter/tests/test_group_create_command_handler.py]: Isn't rider id is integer type?

File: frontend/src/containers/userGroup/UserGroupContainer.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+function UserGroupContainer() {
+  const dispatch = useDispatch();
+  const group = useSelector(group => ({
+    group: group.group,
+  }));
+
+  const onClickOnTaxi = useCallback(id => dispatch(onTaxi(id)), [dispatch]);
+  // group.driver is currently null
+  // so will use temporary driver info
+  const driverInfo = 'Name: mockDriver, Vehicle : BMW, Plate No. : 01A 1234';
```
Comment (zeroFruit) [inline frontend/src/containers/userGroup/UserGroupContainer.js]: `driverInfo` should be object type! Aren't we fixed in last pull request??

Comment (dkim94) [reply frontend/src/containers/userGroup/UserGroupContainer.js]: > `driverInfo` should be object type! Aren't we fixed in last pull request??

This is object on upstream master. I think something went wrong on merging :(

File: backend/api_gateway/group/views.py (hunk 21 above)
Comment (mss8087) [reply backend/api_gateway/group/views.py]: I will check

**Pull request merged**